File: lcp/LCP 30. 魔塔游戏/Solution.cpp
class Solution {
public:
    int magicTower(vector<int>& nums) {
        priority_queue<int, vector<int>, greater<int>> q;
        int64_t blood = 1, v = 0;
        int ans = 0;
        for (int x : nums) {
            if (x < 0) {
                q.push(x);
            }
            blood += x;
            if (blood <= 0) {
                ++ans;
                v += q.top();
                blood -= q.top();
                q.pop();
            }
        }
        blood += v;
        return blood <= 0 ? -1 : ans;
    }
};

File: lcp/LCP 30. 魔塔游戏/Solution.py
class Solution:
    def magicTower(self, nums: List[int]) -> int:
        q = []
        blood = 1
        ans = v = 0
        for x in nums:
            if x < 0:
                heappush(q, x)
            blood += x
            if blood <= 0:
                ans += 1
                v += q[0]
                blood -= heappop(q)
        blood += v
        return -1 if blood <= 0 else ans
